### PR TITLE
Add official support for Python 3.12

### LINKED ISSUE #3287: Add official support for Python 3.12
## Description

Kedro itself probably works on Python 3.12 already, it would be nice to declare official support.

However, installing Kedro is one thing, but installing the typical dependencies might not be straightforward. For example, I just tested the spaceflights starter and most of the dependencies have already published precompiled wheels for Python 3.12 (at least for M1 Mac), but two of them are still problematic as of today:

- aiohttp https://github.com/aio-libs/aiohttp/issues/7739 worked by installing the beta version as advised there, so it will be solved soon ⌛ 
- pyzmq https://github.com/zeromq/pyzmq/issues/1907 (M1 specific), didn't work after installing the ZMQ header libraries with mamba, resolution timeline unclear ❓ 

## Context

#2815 was already completed, but officially Kedro does not support Python 3.12 yet.

You can use Kedro on Python 3.12 by manually disabling the warning.

## Possible Implementation

Wait a bit until at least the spaceflights starter can be safely installed in most mainstream platforms.

## Possible Alternatives

Declare Python 3.12 support already, at the cost of creating some grievance of users that then proceed to install some dependencies.

### PR DESCRIPTION
**Pull request by astrojuanlu**
Close #3287.

## Development notes

First attempt, I haven't ran the test suite locally yet. But `pip install -e .[test]` worked first time on Apple Silicon.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

**Comment by astrojuanlu** [comment]
Tests failing on 3.11 instead 🤔 

```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
    obj = _ForkingPickler.dumps(obj)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
AttributeError: Can't pickle local object 'PluginManager.add_hookcall_monitoring.<locals>.traced_hookexec'
```

rings a bell but I don't have time to look at this right now, will do so later.

**Comment by merelcht** [comment]
> Tests failing on 3.11 instead 🤔
> 
> ```
> Traceback (most recent call last):
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
>     obj = _ForkingPickler.dumps(obj)
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
>     cls(buf, protocol).dump(obj)
> AttributeError: Can't pickle local object 'PluginManager.add_hookcall_monitoring.<locals>.traced_hookexec'
> ```
> 
> rings a bell but I don't have time to look at this right now, will do so later.

This is the flaky test that occasionally fails. I think @lrcouto was seeing this as well on her PR for adding the `_EPHEMERAL` attribute.

**Comment by astrojuanlu** [comment]
```
E               AttributeError: 'called_once_with' is not a valid assertion. Use a spec for the mock if 'called_once_with' is meant to be an attribute.. Did you mean: 'assert_called_once_with'?
```

Interesting 😄

File: tests/config/test_omegaconf_config.py [hunk 1]
```diff
@@ -437,7 +438,8 @@ def test_overlapping_patterns(self, tmp_path, mocker):
 
         mocked_load = mocker.patch("omegaconf.OmegaConf.load")
         expected_path = (tmp_path / "dev" / "user1" / "catalog2.yml").resolve()
-        assert mocked_load.called_once_with(expected_path)
+
+        mocked_load.assert_called_once_with(expected_path)
```
Comment (astrojuanlu) [inline tests/config/test_omegaconf_config.py:442]: See https://github.com/python/cpython/issues/100690

This assert was trivially passing, the test was not doing anything (and it's broken at least in 3.11 and 3.12)

**Comment by astrojuanlu** [comment]
One test failure left, only happening on Python 3.12:

```
    @use_config_dir
    def test_load_core_config_get_syntax(self, tmp_path):
        """Make sure core config can be fetched with .get()"""
        conf = OmegaConfigLoader(
            str(tmp_path), base_env=_BASE_ENV, default_run_env=_DEFAULT_RUN_ENV
        )
        params = conf.get("parameters")
        catalog = conf.get("catalog")
    
>       assert params["param1"] == 1
E       TypeError: 'NoneType' object is not subscriptable
```

**Comment by lrcouto** [comment]
> > Tests failing on 3.11 instead 🤔
> > ```
> > Traceback (most recent call last):
> >   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/queues.py", line 244, in _feed
> >     obj = _ForkingPickler.dumps(obj)
> >           ^^^^^^^^^^^^^^^^^^^^^^^^^^
> >   File "/opt/hostedtoolcache/Python/3.11.7/x64/lib/python3.11/multiprocessing/reduction.py", line 51, in dumps
> >     cls(buf, protocol).dump(obj)
> > AttributeError: Can't pickle local object 'PluginManager.add_hookcall_monitoring.<locals>.traced_hookexec'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > rings a bell but I don't have time to look at this right now, will do so later.
> 
> This is the flaky test that occasionally fails. I think @lrcouto was seeing this as well on her PR for adding the `_EPHEMERAL` attribute.

This test is different than the one I've seen. There was one that was caused by a requirement version, which Nok fixed, and another that happened in the tests for the starters. This one is new to me.

**Comment by ankatiyar** [comment]
> One test failure left, only happening on Python 3.12:
> 
> ```
>     @use_config_dir
>     def test_load_core_config_get_syntax(self, tmp_path):
>         """Make sure core config can be fetched with .get()"""
>         conf = OmegaConfigLoader(
>             str(tmp_path), base_env=_BASE_ENV, default_run_env=_DEFAULT_RUN_ENV
>         )
>         params = conf.get("parameters")
>         catalog = conf.get("catalog")
>     
> >       assert params["param1"] == 1
> E       TypeError: 'NoneType' object is not subscriptable
> ```

Did a little digging - https://github.com/python/cpython/issues/105524

**Comment by astrojuanlu** [comment]
Looks like this will require a bit of work. I'm not pursuing this PR for the time being.

**Comment by merelcht** [comment]
> Looks like this will require a bit of work. I'm not pursuing this PR for the time being.

@astrojuanlu Maybe you can add some context on the challenges you're seeing in https://github.com/kedro-org/kedro/issues/3287 and then we can properly plan it for a sprint?

**Comment by astrojuanlu** [comment]
Good point, done https://github.com/kedro-org/kedro/issues/3287#issuecomment-1956164676

File: tests/config/test_omegaconf_config.py [hunk 2]
```diff
@@ -401,6 +401,7 @@ def test_empty_catalog_file(self, tmp_path):
         )["catalog"]
         assert catalog == {}
 
+    @pytest.mark.xfail(reason="Logic failing")
```
Comment (AhdraMeraliQB) [inline tests/config/test_omegaconf_config.py:404]: Fixed the test to check for the intended behaviour. However, it fails, and manual testing shows that this is due to the implementation of the OmegaConfigLoader. I would suggest correcting this should be split out into a separate issue to be addressed.

File: kedro/config/abstract_config.py [hunk 3]
```diff
@@ -26,6 +26,15 @@ def __init__(
...
 
+    # From Python 3.12 __getitem__ isn't called in UserDict.get()
+    # Use the version from 3.11 and prior
+    def get(self, key: str, default: Any = None) -> Any:
+        "D.get(k[,d]) -> D[k] if k in D, else d.  d defaults to None."
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
```
Comment (noklam) [inline kedro/config/abstract_config.py:37]: Why is this needed?

File: tests/config/test_omegaconf_config.py (hunk 2 above)
Comment (merelcht) [reply tests/config/test_omegaconf_config.py:406]: I don't agree with this being addressed in a separate issue, adding the `3.12` support is what's making this fail and thus it should be fixed in this PR.

File: tests/config/test_omegaconf_config.py [hunk 4]
```diff
@@ -421,23 +424,27 @@ def test_overlapping_patterns(self, tmp_path, mocker):
...
-            "env": "dev",
-            "common": "common",
-            "dev_specific": "wiz",
-            "user1_c2": True,
-        }
-        assert catalog == expected_catalog
+        # Use a mocked function to keep track of function calls
+        with mocker.patch(
+            "omegaconf.OmegaConf.load", wraps=OmegaConf.load
+        ) as mocked_load:
```
[the remark this answers is not in the document]
Comment (AhdraMeraliQB) [reply tests/config/test_omegaconf_config.py]: TIL

File: tests/config/test_omegaconf_config.py [hunk 5]
```diff
@@ -424,27 +424,24 @@ def test_overlapping_patterns(self, tmp_path, mocker):
...
-                config_patterns=catalog_patterns,
-            )["catalog"]
-            expected_catalog = {
-                "env": "dev",
-                "common": "common",
-                "dev_specific": "wiz",
-                "user1_c2": True,
-            }
-            assert catalog == expected_catalog
+        load_spy = mocker.spy(OmegaConf, "load")
```
Comment (astrojuanlu) [inline tests/config/test_omegaconf_config.py:427]: TIL!

File: kedro/config/abstract_config.py (hunk 3 above)
Comment (noklam) [reply kedro/config/abstract_config.py]: Should this be on OmegaConfigLoader or the abstract method? I am slightly confused why `get` method is needed for OmegaConf.

Comment (AhdraMeraliQB) [reply kedro/config/abstract_config.py]: > Should this be on OmegaConfigLoader or the abstract method?

I went for the AbstractConfigLoader to catch any cases where users have defined their own Config Loaders (for whatever reason) that inherit from the AbstractConfigLoader. 

> I am slightly confused why get method is needed for OmegaConf.

Whilst we don't promote using get over the dict syntax, we have made it available - and it features in our API docs - allowing users to do something like:
```
conf_loader = OmegaConfigLoader(<...>)
if not conf_loader.get("parameters"):
    parameters = MY_LOCAL_PARAMS
```

Ultimately, removing get() is a breaking change, which would block adding 3.12 support until the next major release.

File: tests/config/test_omegaconf_config.py (hunk 2 above)
Comment (AhdraMeraliQB) [reply tests/config/test_omegaconf_config.py:406]: It's not the move to include 3.12 that's causing this test to fail, but rather it was never truly passing, the previous method call used for the assert was mistyped, leading to a trivial pass. To resolve this requires digging further into the implementation of the OmegaConfigLoader, as it is not behaving as expected, and perhaps never was. See [xref](https://github.com/kedro-org/kedro/pull/3587#discussion_r1476292680).

**Review by merelcht — APPROVED** [review]
Thanks @astrojuanlu and @AhdraMeraliQB for making this work 👍

**Review by ankatiyar — APPROVED** [review]
Thanks @astrojuanlu & @AhdraMeraliQB :D
